test/main/deepEqual.js

```diff
@@ -146,16 +146,66 @@ QUnit.test( "Objects basics", function( assert ) {
 		}
 	), false );
 
-	// Objects with no prototype, created via Object.create(null), are used e.g. as dictionaries.
+});
+
+QUnit[ typeof Object.create === "function" ? "test" : "skip" ](
+	"Objects with null prototypes", function( assert ) {
+
+	var nonEmptyWithNoProto;
+
+	// Objects with no prototype, created via Object.create(null), are used
+	// e.g. as dictionaries.
 	// Being able to test equivalence against object literals is quite useful.
-	if ( typeof Object.create === "function" ) {
-		assert.equal( QUnit.equiv( Object.create( null ), {} ), true, "empty object without prototype VS empty object" );
+	assert.equal(
+		QUnit.equiv( Object.create( null ), {} ),
+		true,
+		"empty object without prototype VS empty object"
+	);
+
+	assert.equal(
+		QUnit.equiv( {}, Object.create( null ) ),
+		true,
+		"empty object VS empty object without prototype"
+	);
 
-		var nonEmptyWithNoProto = Object.create( null );
-		nonEmptyWithNoProto.foo = "bar";
+	nonEmptyWithNoProto = Object.create( null );
+	nonEmptyWithNoProto.foo = "bar";
 
-		assert.equal( QUnit.equiv( nonEmptyWithNoProto, { foo: "bar" } ), true, "object without prototype VS object" );
-	}
+	assert.equal(
+		QUnit.equiv( nonEmptyWithNoProto, { foo: "bar" } ),
+		true,
+		"object without prototype VS object"
+	);
+
+	assert.equal(
+		QUnit.equiv( { foo: "bar" }, nonEmptyWithNoProto ),
+		true,
+		"object VS object without prototype"
+	);
+});
+
+// Ref #851
+QUnit[ typeof Object.create === "function" ? "test" : "skip" ](
+	"Object prototype constructor is null", function( assert ) {
+
+	// Ref #851
+	// Unfortunately, in practice `Object.create(null)` is fairly costly.
+	// To mitigate this cost a specialized NullObject can be used. This
+	// Object has similar safe characteristics, but with dramatically
+	// reduced allocation costs.
+	function NullObject() {}
+	NullObject.prototype = Object.create( null, {
+		constructor: {
+			value: null
+		}
+	});
+
+	var a = new NullObject();
+	a.foo = 1;
+	var b = { foo: 1 };
+
+	assert.ok( QUnit.equiv( a, b ) );
+	assert.ok( QUnit.equiv( b, a ) );
```

Comment:
@stefanpenner 

Also:

var c = Object.create( a );

c.constructor === null, so maybe the native chain walk also answers your question, right?